### PR TITLE
[WIP] initial tests for key backup

### PR DESCRIPTION
**Pull request by uhoreg**
tests for https://github.com/matrix-org/synapse/pull/2731 (or whatever its successor will end up being)

### CODE REVIEW

File: tests/41end-to-end-keys/07-backup.pl [hunk 1]
```diff
@@ -0,0 +1,271 @@
+my $fixture = local_user_fixture();
+
+my $current_version; # FIXME: is there a better way of passing the backup version between tests?
```
Comment (ara4n) [inline tests/41end-to-end-keys/07-backup.pl]: this is fine.

File: tests/41end-to-end-keys/07-backup.pl [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+   proves => [qw( can_create_backup_version )],
+
+   do => sub {
+      my ( $user ) = @_;
+
+      do_request_json_for( $user,
+         method  => "POST",
+         uri     => "/unstable/room_keys/version",
+         content => {
+            algorithm => "m.megolm_backup.v1",
```
Comment (ara4n) [inline tests/41end-to-end-keys/07-backup.pl]: hm, shouldn't we use a real algorithm for this? (or is this a real one? it surprises me it doesn't include info about the ciphers being used, like the m.room.encryption algorithm does)

File: tests/41end-to-end-keys/07-backup.pl [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+         assert_json_keys( $content, "auth_data" );
+
+         $content->{algorithm} eq "m.megolm_backup.v1" or
+            die "Expected algorithm to match submitted data";
+
+         $content->{auth_data} eq "anopaquestring" or
+            die "Expected auth_data to match submitted data";
+
+         # FIXME: check that version matches the version returned above
```
Comment (ara4n) [inline tests/41end-to-end-keys/07-backup.pl]: hm, isn't this just comparing $current_version with $content->{version} or something?

File: tests/41end-to-end-keys/07-backup.pl [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+         );
+      })->then( sub {
+         my ( $content ) = @_;
+         log_if_fail "Content", $content;
+
+         assert_json_keys( $content, "algorithm" );
+
+         assert_json_keys( $content, "auth_data" );
+
+         $content->{algorithm} eq "m.megolm_backup.v1" or
```
Comment (ara4n) [inline tests/41end-to-end-keys/07-backup.pl]: nitpicking, but i'd probably factor out these constants in the name of DRY

Comment (ara4n) [reply tests/41end-to-end-keys/07-backup.pl]: actually, ignore this, it's more legible at it is.

File: tests/41end-to-end-keys/07-backup.pl [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+         assert_json_keys( $content, "is_verified" );
+
+         assert_json_keys( $content, "session_data" );
+
+         # The data should not be overwritten, so should be the same as what
+         # was set by the previous test.
+         $content->{first_message_index} == 1 or
+            die "Expected first message index to match submitted data";
+
+         $content->{forwarded_count} == 0 or
```
Comment (ara4n) [inline tests/41end-to-end-keys/07-backup.pl]: we should probably exercise `forwarded_count` here too - perhaps just put another update attempt in here to check that forwarded keys don't trump non-forwarded ones? (iirc that's how the logic went)

**Comment by ara4n** [comment]
lgtm, other than minor feedback!

**Comment by uhoreg** [comment]
Reminder to myself before I forget: still need to check the bulk uploading/downloading endpoints, as well as the delete endpoints. Also the the get endpoints without a version specified

File: tests/41end-to-end-keys/07-backup.pl (hunk 3 above)
Comment (uhoreg) [reply tests/41end-to-end-keys/07-backup.pl]: Yup, except that I don't think the version is returned yet, so I need to fix synapse to do that first. ;)

File: tests/41end-to-end-keys/07-backup.pl (hunk 2 above)
Comment (uhoreg) [reply tests/41end-to-end-keys/07-backup.pl]: That's the algorithm given in the proposal, but I agree, it should probably include some algorithm info

File: tests/41end-to-end-keys/07-backup.pl (hunk 3 above)
Comment (uhoreg) [reply tests/41end-to-end-keys/07-backup.pl]: So it turns out that Synapse *does* return the version, even though it wasn't in the proposal. :man_shrugging:

Comment (ara4n) [reply tests/41end-to-end-keys/07-backup.pl]: oops. i can imagine that i fixed up some of the thinkos in the proposal whilst doing the implementation, but then given the implementation never got fully finished, the proposal wasn't updated - sorry.

**Comment by dbkr** [comment]
I've added more tests in https://github.com/matrix-org/sytest/pull/486

**Comment by dbkr** [comment]
Superseded by https://github.com/matrix-org/sytest/pull/503

**Pull request closed without merging**